api/v1alpha1/nodehealthcheck_webhook.go

```diff
@@ -74,76 +79,126 @@ func (r *NodeHealthCheck) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	if nhc.Spec.MinHealthy.Type == intstr.Int && nhc.Spec.MinHealthy.IntVal < 0 {
+		return fmt.Errorf("%s: %v", minHealthyError, nhc.Spec.MinHealthy)
 	}
 	return nil
 }
 
-func (r *NodeHealthCheck) validateSelector() error {
-	if _, err := metav1.LabelSelectorAsSelector(&r.Spec.Selector); err != nil {
+func (nhc *NodeHealthCheck) validateSelector() error {
+	if _, err := metav1.LabelSelectorAsSelector(&nhc.Spec.Selector); err != nil {
 		return fmt.Errorf("%s: %v", invalidSelectorError, err.Error())
 	}
 	return nil
 }
 
-func (r *NodeHealthCheck) isRestrictedFieldUpdated(old *NodeHealthCheck) (bool, string) {
+func (nhc *NodeHealthCheck) validateMutualRemediations() error {
+	if nhc.Spec.RemediationTemplate == nil && (nhc.Spec.EscalatingRemediations == nil || len(nhc.Spec.EscalatingRemediations) == 0) {
+		return fmt.Errorf(mandatoryRemediationError)
+	}
+	if !(nhc.Spec.RemediationTemplate == nil || nhc.Spec.EscalatingRemediations == nil) {
+		return fmt.Errorf(mutualRemediationError)
+	}
+	return nil
+}
+
+func (nhc *NodeHealthCheck) validateEscalatingRemediations() error {
+	if nhc.Spec.EscalatingRemediations == nil {
+		return nil
+	}
+
+	aggregated := errors.NewAggregate([]error{
+		nhc.validateEscalatingRemediationsUniqueOrder(),
+		nhc.validateEscalatingRemediationsTimeout(),
+	})
+	return aggregated
+}
+
+func (nhc *NodeHealthCheck) validateEscalatingRemediationsUniqueOrder() error {
+	orders := make(map[int]struct{}, len(nhc.Spec.EscalatingRemediations))
+	for _, rem := range nhc.Spec.EscalatingRemediations {
+		if _, exists := orders[rem.Order]; exists {
+			return fmt.Errorf("%s: found duplicate order %v", uniqueOrderError, rem.Order)
+		}
+		orders[rem.Order] = struct{}{}
+	}
+	return nil
+}
+
+func (nhc *NodeHealthCheck) validateEscalatingRemediationsTimeout() error {
+	for _, rem := range nhc.Spec.EscalatingRemediations {
+		if rem.Timeout.Duration < 1*time.Minute {
+			return fmt.Errorf("%s: found timeout %v", minimumTimeoutError, rem.Timeout)
+		}
+	}
+	return nil
+}
+
+func (nhc *NodeHealthCheck) isRestrictedFieldUpdated(old *NodeHealthCheck) (bool, string) {
 	// modifying these fields can cause dangling remediations
-	if !reflect.DeepEqual(r.Spec.Selector, old.Spec.Selector) {
+	if !reflect.DeepEqual(nhc.Spec.Selector, old.Spec.Selector) {
 		return true, "selector"
 	}
-	if !reflect.DeepEqual(r.Spec.RemediationTemplate, old.Spec.RemediationTemplate) {
+	if !reflect.DeepEqual(nhc.Spec.RemediationTemplate, old.Spec.RemediationTemplate) {
 		return true, "remediation template"
 	}
+	if !reflect.DeepEqual(nhc.Spec.EscalatingRemediations, old.Spec.EscalatingRemediations) {
+		return true, "escalating remediations"
+	}
 	return false, ""
 }
 
-func (r *NodeHealthCheck) isRemediating() bool {
-	return len(r.Status.InFlightRemediations) > 0
+func (nhc *NodeHealthCheck) isRemediating() bool {
+	return len(nhc.Status.InFlightRemediations) > 0 || len(nhc.Status.UnhealthyNodes) > 0
```

Comment:
Nit: IIUC `UnhealthyNodes` would replace `InFlightRemediations`, so the first check is redundant.

Reply:
we can have old NHCs with ongoing remediation where the new status isn't updated yet

Reply:
Do you mean in case of an upgrade were there are already existing remediations ?